pf/proto/protov6_test.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proto_test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/hexops/autogold/v2"
+	"github.com/stretchr/testify/require"
+
+	"github.com/pulumi/pulumi-terraform-bridge/pf/proto"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge/info"
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+)
+
+func marshalProviderShim(t *testing.T, p shim.Provider) []byte {
+	var b bytes.Buffer
+	enc := json.NewEncoder(&b)
+	enc.SetEscapeHTML(false)
+	enc.SetIndent("", "    ")
+	require.NoError(t, enc.Encode(info.MarshalProviderShim(p)))
+	return b.Bytes()
+}
+
+func TestShimSchema(t *testing.T) {
+	autogold.Expect("{}\n").Equal(t, string(marshalProviderShim(t, proto.Empty())))
+}
+
+func TestDynamicType(t *testing.T) {
+	b := marshalProviderShim(t,
+		proto.New(context.Background(), providerServer{
+			SchemaResponse: &tfprotov6.GetProviderSchemaResponse{
+				ResourceSchemas: map[string]*tfprotov6.Schema{
+					"my_res": {Block: &tfprotov6.SchemaBlock{
+						Attributes: []*tfprotov6.SchemaAttribute{{
+							Type: tftypes.DynamicPseudoType,
+						}},
+					}},
+				},
+			},
+		}))
+	autogold.Expect(`{
+    "resources": {
+        "my_res": {
+            "": {
```

Comment:
I'm a bit confused by the schema here?
Why is this ""? Also isn't it supposed to error - why is this generated?

Reply:
It's `""` because I didn't pass a `Name` value to `tfprotov6.SchemaAttribute`. I changed the commit to make this clearer.

Reply:
If I set up an integration test, then it would error. The behavior under test is working correctly. `8` is the value for a `shim.TypeDynamic`, which is correct:

https://github.com/pulumi/pulumi-terraform-bridge/blob/d609089b6e4a014e49f2d54280b4c38e2172b5ef/pkg/tfshim/shim.go#L64

If I tried to convert this `shim.Provider` into a Pulumi schema, then it would error here:

https://github.com/pulumi/pulumi-terraform-bridge/blob/d609089b6e4a014e49f2d54280b4c38e2172b5ef/pkg/tfgen/generate.go#L439-L440

The error is already under test here:

https://github.com/pulumi/pulumi-terraform-bridge/blob/d609089b6e4a014e49f2d54280b4c38e2172b5ef/pkg/tfgen/generate_test.go#L172-L177